src/Node/Message/Store.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE OverloadedStrings     #-}
+
+module Node.Message.Store
+    ( StoreP (..)
+    ) where
+
+import qualified Data.Store                    as Store
+import qualified Data.ByteString               as BS
+import qualified Data.ByteString.Lazy          as LBS
+import           Data.Word                     (Word32)
+import qualified Network.Transport.Internal    as NT (encodeWord32, decodeWord32)
+import           Node.Message.Class            (Serializable (..))
+import           Node.Message.Decoder          (Decoder (..))
+
+data StoreP = StoreP
+
+instance Store.Store t => Serializable StoreP t where
+
+    -- Length-prefix the store-encoded body. The length is assumed to fit into
+    -- 32 bits.
+    packMsg _ t = encoded
+      where
+        encodedBody = Store.encode t
+        encodedLength = NT.encodeWord32 (fromIntegral (BS.length encodedBody))
+        encoded = LBS.fromStrict (BS.append encodedLength encodedBody)
+
+    unpackMsg _ = storeDecoder BS.empty
+
+storeDecoder :: Store.Store t => BS.ByteString -> Decoder t
+storeDecoder bs = Partial $ \mbs -> case mbs of
+    Nothing -> Fail BS.empty (fromIntegral (BS.length bs)) "Unexpected end of input (length prefix)"
+    Just bs' ->
+        let (front, back) = BS.splitAt 4 (BS.append bs bs')
+        in  if BS.length front == 4
+            then storeDecoderBody (NT.decodeWord32 front) BS.empty (Just back)
+            else storeDecoder front
```

Comment:
The whole observed input so far `BS.append bs bs'` is split at the 4th byte. If the front isn't of length 4, then it must be shorter than 4 bytes, so we need to wait for more input. In this case, `back` is empty so we can forget about it.